pkg/trie/triedb/node.go

```diff
@@ -26,11 +30,48 @@ type (
 	}
 
 	newValueRef struct {
-		hash *common.Hash
+		hash common.Hash
 		Data []byte
 	}
 )
 
+func NewEncodedValue(value nodeValue, partial []byte, childF onChildStoreFn) (codec.EncodedValue, error) {
+	switch v := value.(type) {
+	case inline:
+		return codec.NewInlineValue(v.Data), nil
+	case valueRef:
+		return codec.NewHashedValue(v.hash.ToBytes()), nil
+	case newValueRef:
+		// Store value in db
+		childRef, err := childF(NewNodeToEncode{partialKey: partial, value: v.Data}, partial, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// Check and get new new value hash
+		switch cr := childRef.(type) {
+		case HashChildReference:
+			if cr.hash == common.EmptyHash {
+				panic("new external value are always added before encoding a node")
+			}
+
+			if v.hash != common.EmptyHash {
+				if v.hash != cr.hash {
+					panic("hash mismatch")
+				}
+			}
```

Comment:
```suggestion
			if v.hash != common.EmptyHash && v.hash != cr.hash {
					panic("hash mismatch")
			}
```

Reply:
There was an else missing, [this is the logic](https://github.com/paritytech/trie/blob/c3348e4b335443c17586a8b7491ba893cfde76a6/trie-db/src/triedbmut.rs#L166-L170) in trie crate.
I added it and this will also solves your [second](https://github.com/ChainSafe/gossamer/pull/4039/files#r1653624350) comment